feat: remove delegator address in WrappedFilePV, change init cmd and add migration cmd

wnjoon opened the pull request:
This PR includes the following changes:

**1. Remove delegator address in WrappedFilePV:**

- Removed Delegator Address field from WrappedFilePV. Delegator Address can be obtained from kvStore of x/checkpointing module via bls public key.
- Simplify the structure of WrappedFilePV.

**2. Addition of InitCmd**:

- InitCmd has been updated to generate both Comet and BLS keys through babylond init.
- The --bls-password flag allows you to specify a BLS password. If the flag is not provided, the password will be requested interactively via a prompt.

**3. BLS Key File Migration**:

- The babylond migrate command allows the priv_validator_key.json file from previous versions to be converted into the separated key file structure used in the current version.
- If the priv_validator_key.json file does not exist or already contains only Comet key information (indicating it is in the latest format), no migration will occur.

**4. Fix for `e2e-run-upgrade-v1` Error**:

- Resolved an issue in e2e-run-upgrade-v1 during e2e testing caused by differences in the key file structures between the current version and the images pulled from Docker Hub for pre-upgrade containers.
- Introduced a new build tag, e2e_upgrade, applied when building the e2e Docker image. This ensures that `babylond start` checks the priv_validator_key.json file, and if it is in an older format, it is automatically migrated.

And we suggest a new discussion about lightweight BlsSigner interface with remove WrappedFilePV structure

- When BLS public key is known, the validator address required for Vote extensions can be obtained using the kvStore of the x/checkpointing module.
- Therefore, the BlsSigner interface can be fully implemented in BlsPVKey, and the WrappedFilePV structure is not needed.
- In cases where CometBFT uses a PrivValidator that does not rely on FilePV (e.g., Horcrux or other SignerClient), the FilePV will not be present. Note that remote signers are currently not supported.
- [comment for reference](https://github.com/b-harvest/babylon/pull/8#discussion_r1921791859)
- Please consider whether this proposal is necessary and give us your opinion.

gitferry commented on cmd/babylond/cmd/migrate.go:
```diff
@@ -0,0 +1,124 @@
+	cmd := &cobra.Command{
+		Use:   "migrate-bls-key",
+		Short: "Migrate the contents of the priv_validator_key.json file into separate files of bls and comet",
+		Long: strings.TrimSpace(`migrate splits the contents of the priv_validator_key.json file, 
+		which contained both the bls and comet keys used in previous versions, into separate files.
+
+BLS keys are stored along with other validator keys in priv_validator_key.json in previous version,
+which should exist before running the command (via babylond init or babylond testnet).
+
+NOTE: Before proceeding with the migration, ensure you back up the priv_validator_key.json file to a secure location.
+This will help prevent potential loss of critical validator information in case of issues during the migration process.
+
+Example:
+$ babylond migrate-bls-key --home ./
+`,
```
```suggestion
		Long: strings.TrimSpace(`Migration splits the contents of the priv_validator_key.json file, 
		which contained both the bls and comet keys used in previous versions, into separate files.

BLS keys are stored along with the Ed25519 validator key in priv_validator_key.json in the previous version,
which should exist before running the command (via babylond init or babylond testnet).

NOTE: Before proceeding with the migration, ensure you back up the priv_validator_key.json file to a secure location.
This will help prevent potential loss of critical validator information in case of issues during the migration process.

Example:
$ babylond migrate-bls-key --home ./
`,
```

wnjoon replied:
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

gitferry commented on cmd/babylond/cmd/migrate.go:
migrate.go -> migrate_bls.go

wnjoon replied:
How about changing it to migrate_bls_key.go for consistency with create_bls_key.go?

wnjoon replied:
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

gitferry commented on cmd/babylond/cmd/migrate.go:
```diff
@@ -0,0 +1,124 @@
+	filepath := cmtcfg.PrivValidatorKeyFile()
+
+	if !cmtos.FileExists(filepath) {
+		return fmt.Errorf("priv_validator_key.json of previous version not found")
```
```suggestion
		return fmt.Errorf("priv_validator_key.json of previous version not found in %s", filepath)
```

wnjoon replied:
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

gitferry commented on cmd/babylond/cmd/migrate.go:
```diff
@@ -0,0 +1,124 @@
+
+	pv, err := loadPrevWrappedFilePV(filepath)
+	if err != nil {
+		return err
```
let's extend the error message

wnjoon replied:
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

wnjoon replied:
@gitferry 
In addition, there are some errs in the code that are not extended, other than the code we changed. First, we fixed the contents that are not extended in the code we wrote. We will fix the rest as we find them while working on the code.

gitferry commented on cmd/babylond/cmd/migrate.go:
```diff
@@ -0,0 +1,124 @@
+	blsPv := privval.NewBlsPV(prevBlsPrivKey, privval.DefaultBlsKeyFile(homeDir), privval.DefaultBlsPasswordFile(homeDir))
+
+	// before saving keys to files, verify that the migrated keys match
+	if err := verifyAfterMigration(
```
I think this is `verifyBeforeMigration` other than `verifyAfterMigration`. I think we should read keys from speparated `priv_validator_key.json` and `bls_key.json` and verify

wnjoon replied:
What you said is that I have to save the separated key to each file and then read the contents of each file again to verify it?

I thought that I should save it to the file first and then verify it if it fails, so I tried to verify it with the data I want to save to the file and then save it to the file. What do you think?

gitferry replied:
> I thought that I should save it to the file first and then verify it if it fails, so I tried to verify it with the data I want to save to the file and then save it to the file. What do you think?

I think we just need to verify it after saving by reading the file, no need to verify it before saving

wnjoon replied:
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

If the meaning of Migration is the process of separating an existing file that was one into two, then the function name can also be seen as `verifyAfterMigration`. However, if the completion of the migrate function means the end of Migration, then `verifyBeforeMigration` is correct.

So, to remove any confusion about the meaning, I named it `verifySeparateFiles`. If you have any other opinions, please let me know.

gitferry commented on cmd/babylond/cmd/migrate.go:
```diff
@@ -0,0 +1,124 @@
+)
+
+// PrevWrappedFilePV is a struct for prev version of priv_validator_key.json
+type PrevWrappedFilePV struct {
```
Isn't previous wrapped file pv something like this https://github.com/babylonlabs-io/babylon/blob/5f8af8ced17d24f3f0c6172293cd37fb3d055807/privval/file.go#L49?

wnjoon replied:
That's right. However, since migration only needs private keys of both the comet and the bls, I created a PrevWrappedFilePV structure that only contains the data corresponding to unmarshalling.

gitferry commented on privval/file.go:
```diff
@@ -1,116 +1,45 @@
+	cmtcrypto "github.com/cometbft/cometbft/crypto"
+	cmtprivval "github.com/cometbft/cometbft/privval"
 )
 
```
let's add `var _ keeper.BlsSigner = &WrappedFilePV{}`

wnjoon replied:
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

gitferry commented on testutil/signer/private.go:
```diff
@@ -34,8 +35,12 @@ func SetupTestPrivSigner() (*signer.PrivSigner, error) {
 
-func GenesisKeyFromPrivSigner(ps *signer.PrivSigner) (*checkpointingtypes.GenesisKey, error) {
-	valKeys, err := privval.NewValidatorKeys(ps.WrappedPV.GetValPrivKey(), ps.WrappedPV.GetBlsPrivKey())
+// func GenesisKeyFromPrivSigner(ps *signer.PrivSigner, delegatorAddress sdk.ValAddress) (*checkpointingtypes.GenesisKey, error) {
```
leftover?

wnjoon replied:
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

gitferry commented on x/checkpointing/keeper/bls_signer.go:
```diff
@@ -9,10 +9,10 @@ import (
 
 type BlsSigner interface {
-	GetAddress() sdk.ValAddress
+	// GetAddress() sdk.ValAddress
```
leftover?

wnjoon replied:
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

gitferry commented on x/checkpointing/vote_ext.go:
```diff
@@ -53,15 +53,17 @@ func (h *VoteExtensionHandler) ExtendVote() sdk.ExtendVoteHandler {
-		_, _, err := curValSet.FindValidatorWithIndex(signer)
+
+		valConsAddr := k.GetConAddressFromPubkey()
+		val, err := k.GetValidatorByConsAddr(ctx, valConsAddr)
```
I wondered if we need this check at all, given that `ExtendVote` is only called if the validator is in the validator set. Wdyt? @KonradStaniec

KonradStaniec replied:
I would leave it tbh it is not a big cost, and if something change on cometbft side (or they will be some bug) we will quickly catch it after upgrade

wnjoon replied:
https://github.com/babylonlabs-io/babylon/pull/434#discussion_r1926493457

If what you said is true, there seems to be no reason for this content to be immediately modified in the current PR.

gitferry commented on x/checkpointing/vote_ext_test.go:
```diff
@@ -213,40 +213,43 @@ func FuzzExtendVote_EmptyBLSPrivKey(f *testing.F) {
+// WARN
+// This test code became untestable after deleting the DelegatorAddress field.
+//
+// // FuzzExtendVote_NotInValidatorSet tests the case where the
+// // private signer is not in the validator set
+// func FuzzExtendVote_NotInValidatorSet(f *testing.F) {
+// 	datagen.AddRandomSeedsToFuzzer(f, 10)
+
+// 	f.Fuzz(func(t *testing.T, seed int64) {
+// 		r := rand.New(rand.NewSource(seed))
+// 		// generate the validator set with 10 validators as genesis
+// 		genesisValSet, ps, err := datagen.GenesisValidatorSetWithPrivSigner(10)
+// 		require.NoError(t, err)
+
+// 		// the private signer is not included in the validator set
+// 		helper := testhelper.NewHelperWithValSetNoSigner(t, genesisValSet, ps)
+
+// 		ek := helper.App.EpochingKeeper
+
+// 		epoch := ek.GetEpoch(helper.Ctx)
+// 		require.Equal(t, uint64(1), epoch.EpochNumber)
+
+// 		// go to block 10, reaching epoch boundary
+// 		interval := ek.GetParams(helper.Ctx).EpochInterval
+// 		for i := uint64(0); i < interval-2; i++ {
+// 			_, err := helper.ApplyEmptyBlockWithSomeInvalidVoteExtensions(r)
+// 			require.NoError(t, err)
+// 		}
+
+// 		req := &abci.RequestExtendVote{
+// 			Hash:   datagen.GenRandomByteArray(r, types.HashSize),
+// 			Height: 10,
+// 		}
+
+// 		// error is expected because the BLS signer in not
+// 		// in the validator set
+// 		_, err = helper.App.ExtendVote(helper.Ctx, req)
+// 		require.Error(t, err)
+// 	})
+// }
```
Think we can remove it

wnjoon replied:
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

gitferry reviewed (commented):
In general looks good to me! Let's wait approval from @KonradStaniec. Some comments:

gitferry commented:
> And we suggest a new discussion about lightweight BlsSigner interface with remove WrappedFilePV structure

Totally agree with the suggestion! For bls signing, the comet pv is not needed. `WrappedFilePV` is legacy code from where we embeded bls key into `priv_validator_key.json`. This improvement can be done in a separate PR though

KonradStaniec reviewed (approved):
apart from some leftover commented code, lgtm!

wnjoon commented on cmd/babylond/cmd/init.go:
```diff
@@ -0,0 +1,32 @@
+		},
+	}
+	cmd.Flags().AddFlagSet(cosmosInitCmd.Flags())
+	cmd.Flags().String(flagBlsPassword, "", "The password for the BLS key. If a flag is set, the non-empty password should be provided. If a flag is not set, the password will be read from the prompt.")
```
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

wnjoon commented on cmd/babylond/cmd/init.go:
```diff
@@ -0,0 +1,32 @@
+	cmd := &cobra.Command{
+		Use:   cosmosInitCmd.Use,
+		Short: cosmosInitCmd.Short,
+		Long:  cosmosInitCmd.Long,
```
fixed [bbe2efd](https://github.com/babylonlabs-io/babylon/pull/434/commits/bbe2efd813585a1e8cdcb2ec431ca1be8595866c)

wnjoon commented:
Commit Summary

- Changed the logic in migrate-bls-key cmd to load and verify keys stored in separate files. 
- Removed unnecessary comments and added cmd description.
- Error Extension and etc
